src/Features/CSharp/Portable/OrderModifiers/CSharpOrderModifiersHelper.cs

```diff
@@ -0,0 +1,78 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading;
+
+namespace Microsoft.CodeAnalysis.CSharp.OrderModifiers
+{
+    internal static class CSharpOrderModifiersHelper
+    {
+        private static readonly char[] s_comma = { ',' };
+
+        /// <remarks>
+        /// Reference type so we can read/write atomically.
+        /// </remarks>
+        private static Tuple<string, Dictionary<int, int>> s_lastParsed;
+
+        public static bool TryGetOrComputePreferredOrder(string value, out Dictionary<int, int> preferredOrder)
```

Comment:
`ReadOnlyDictionary<TKey, TValue>` has O(1) lookups, where `ImmutableDictionary<TKey, TValue>` has O(log n) lookups. I suggested the former on the basis that this mattered. Higher-performing alternatives which do not compromise data integrity are available in the event this proves to be a performance issue (with measurements).